src/trace_writer.ml

```diff
@@ -875,10 +883,34 @@ let maybe_stop_filtered_region t ~should_write =
     t.in_filtered_region <- false)
 ;;
 
+let write_event_and_callstack events_writer event callstack =
+  let compression_event =
+    Callstack_compression.compress_callstack
+      Tracing_tool_output.(events_writer.callstack_compression_state)
+      (Callstack.(callstack.stack)
+      |> Stack.to_list
+      |> List.map ~f:(fun Event.Location.{ symbol; _ } -> symbol))
+  in
+  let event_and_callstack =
+    Event_and_callstack.{ event; callstack = compression_event }
+  in
+  match events_writer with
+  | Tracing_tool_output.{ format = Sexp; writer; _ } ->
+    Async.Writer.write_sexp
+      ~terminate_with:Async.Writer.Terminate_with.Newline
+      writer
+      [%sexp (event_and_callstack : Event_and_callstack.t)]
+  | Tracing_tool_output.{ format = Binio; writer; _ } ->
+    Async.Writer.write_bin_prot
+      writer
+      Event_and_callstack.bin_writer_t
+      event_and_callstack
+;;
+
 (* Write perf_events into a file as a Fuschia trace (stack events). Events should be
    collected with --itrace=be or cre, and -F pid,tid,time,flags,addr,sym,symoff as per
    the constants defined above. *)
-let write_event (T t) event =
+let write_event (T t) ?events_writer event =
```

Comment:
The diff to this function is pretty hard to read. Can you explain in words what you changed here in a reply to this comment?

Reply:
Huh, I'm really surprised GIT shows the diff so poorly, I guess its because of the changed indentation. All that's changed here is that `event` is being matched on in two steps instead of one. Now it is first matched with `Ok event_value` or `Error ...`, then `write_event_and_callstack` is run, and then `event_value` is being matched further.

Reply:
That makes sense, thanks.